File: plugins/plugin-secrets-manager/rust/src/storage.rs
//! Storage interfaces and implementations for secrets.
//!
//! Provides abstract storage interface and in-memory implementation.

use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;

use crate::types::{
    SecretConfig, SecretContext, SecretLevel, SecretMetadata, SecretsError, SecretsResult,
    StorageBackend,
};

/// Storage entry containing value and configuration.
#[derive(Debug, Clone)]
pub struct StorageEntry {
    /// The stored value (may be encrypted)
    pub value: serde_json::Value,
    /// Configuration for this secret
    pub config: SecretConfig,
}

/// Abstract interface for secret storage.
#[async_trait]
pub trait SecretStorage: Send + Sync {
    /// Get the storage backend type.
    fn storage_type(&self) -> StorageBackend;

    /// Initialize the storage.
    async fn initialize(&self) -> SecretsResult<()>;

    /// Check if a secret exists.
    async fn exists(&self, key: &str, context: &SecretContext) -> SecretsResult<bool>;

    /// Get a secret value.
    async fn get(&self, key: &str, context: &SecretContext) -> SecretsResult<Option<serde_json::Value>>;

    /// Set a secret value.
    async fn set(
        &self,
        key: &str,
        value: serde_json::Value,
        context: &SecretContext,
        config: Option<SecretConfig>,
    ) -> SecretsResult<bool>;

    /// Delete a secret.
    async fn delete(&self, key: &str, context: &SecretContext) -> SecretsResult<bool>;

    /// List secrets (metadata only).
    async fn list(&self, context: &SecretContext) -> SecretsResult<SecretMetadata>;

    /// Get secret configuration.
    async fn get_config(&self, key: &str, context: &SecretContext) -> SecretsResult<Option<SecretConfig>>;

    /// Update secret configuration.
    async fn update_config(
        &self,
        key: &str,
        context: &SecretContext,
        config: SecretConfig,
    ) -> SecretsResult<bool>;
}

/// Generate a storage key from context.
fn make_storage_key(key: &str, context: &SecretContext) -> String {
    match context.level {
        SecretLevel::Global => format!("global:{}", key),
        SecretLevel::World => {
            let world_id = context.world_id.as_deref().unwrap_or("default");
            format!("world:{}:{}", world_id, key)
        }
        SecretLevel::User => {
            let user_id = context.user_id.as_deref().unwrap_or("default");
            format!("user:{}:{}", user_id, key)
        }
    }
}

/// Create default secret configuration.
fn default_config() -> SecretConfig {
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0);

    SecretConfig {
        created_at: Some(now),
        updated_at: Some(now),
        ..Default::default()
    }
}

/// In-memory storage implementation.
pub struct MemorySecretStorage {
    data: Arc<RwLock<HashMap<String, StorageEntry>>>,
}

impl MemorySecretStorage {
    /// Create a new memory storage.
    pub fn new() -> Self {
        Self {
            data: Arc::new(RwLock::new(HashMap::new())),
        }
    }
}

impl Default for MemorySecretStorage {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl SecretStorage for MemorySecretStorage {
    fn storage_type(&self) -> StorageBackend {
        StorageBackend::Memory
    }

    async fn initialize(&self) -> SecretsResult<()> {
        Ok(())
    }

    async fn exists(&self, key: &str, context: &SecretContext) -> SecretsResult<bool> {
        let storage_key = make_storage_key(key, context);
        let data = self.data.read().await;
        Ok(data.contains_key(&storage_key))
    }

    async fn get(&self, key: &str, context: &SecretContext) -> SecretsResult<Option<serde_json::Value>> {
        let storage_key = make_storage_key(key, context);
        let data = self.data.read().await;
        Ok(data.get(&storage_key).map(|e| e.value.clone()))
    }

    async fn set(
        &self,
        key: &str,
        value: serde_json::Value,
        context: &SecretContext,
        config: Option<SecretConfig>,
    ) -> SecretsResult<bool> {
        let storage_key = make_storage_key(key, context);
        let mut data = self.data.write().await;

        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_millis() as i64)
            .unwrap_or(0);

        let existing_config = data.get(&storage_key).map(|e| e.config.clone());
        let mut final_config = config.unwrap_or_else(|| existing_config.unwrap_or_else(default_config));
        final_config.updated_at = Some(now);
        if final_config.created_at.is_none() {
            final_config.created_at = Some(now);
        }

        data.insert(
            storage_key,
            StorageEntry {
                value,
                config: final_config,
            },
        );

        Ok(true)
    }

    async fn delete(&self, key: &str, context: &SecretContext) -> SecretsResult<bool> {
        let storage_key = make_storage_key(key, context);
        let mut data = self.data.write().await;
        Ok(data.remove(&storage_key).is_some())
    }

    async fn list(&self, context: &SecretContext) -> SecretsResult<SecretMetadata> {
        let prefix = match context.level {
            SecretLevel::Global => "global:".to_string(),
            SecretLevel::World => {
                let world_id = context.world_id.as_deref().unwrap_or("default");
                format!("world:{}:", world_id)
            }
            SecretLevel::User => {
                let user_id = context.user_id.as_deref().unwrap_or("default");
                format!("user:{}:", user_id)
            }
        };

        let data = self.data.read().await;
        let mut keys = Vec::new();
        let mut configs = HashMap::new();
        let mut last_modified: Option<i64> = None;

        for (storage_key, entry) in data.iter() {
            if storage_key.starts_with(&prefix) {
                let key = storage_key.strip_prefix(&prefix).unwrap_or(storage_key);
                keys.push(key.to_string());
                configs.insert(key.to_string(), entry.config.clone());

                if let Some(updated) = entry.config.updated_at {
                    last_modified = Some(last_modified.map_or(updated, |lm| lm.max(updated)));
                }
            }
        }

        let count = keys.len();

        Ok(SecretMetadata {
            keys,
            configs,
            count,
            last_modified,
        })
    }

    async fn get_config(&self, key: &str, context: &SecretContext) -> SecretsResult<Option<SecretConfig>> {
        let storage_key = make_storage_key(key, context);
        let data = self.data.read().await;
        Ok(data.get(&storage_key).map(|e| e.config.clone()))
    }

    async fn update_config(
        &self,
        key: &str,
        context: &SecretContext,
        config: SecretConfig,
    ) -> SecretsResult<bool> {
        let storage_key = make_storage_key(key, context);
        let mut data = self.data.write().await;

        if let Some(entry) = data.get_mut(&storage_key) {
            let now = std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|d| d.as_millis() as i64)
                .unwrap_or(0);

            let mut updated_config = config;
            updated_config.updated_at = Some(now);
            updated_config.created_at = entry.config.created_at;
            entry.config = updated_config;
            Ok(true)
        } else {
            Ok(false)
        }
    }
}

/// Composite storage that delegates to different backends based on level.
pub struct CompositeSecretStorage {
    global_storage: Arc<dyn SecretStorage>,
    world_storage: Arc<dyn SecretStorage>,
    user_storage: Arc<dyn SecretStorage>,
}

impl CompositeSecretStorage {
    /// Create a new composite storage.
    pub fn new(
        global_storage: Arc<dyn SecretStorage>,
        world_storage: Arc<dyn SecretStorage>,
        user_storage: Arc<dyn SecretStorage>,
    ) -> Self {
        Self {
            global_storage,
            world_storage,
            user_storage,
        }
    }

    /// Create with default memory storage for all levels.
    pub fn with_memory_storage() -> Self {
        Self {
            global_storage: Arc::new(MemorySecretStorage::new()),
            world_storage: Arc::new(MemorySecretStorage::new()),
            user_storage: Arc::new(MemorySecretStorage::new()),
        }
    }

    /// Get the storage for a given context.
    fn storage_for(&self, context: &SecretContext) -> &Arc<dyn SecretStorage> {
        match context.level {
            SecretLevel::Global => &self.global_storage,
            SecretLevel::World => &self.world_storage,
            SecretLevel::User => &self.user_storage,
        }
    }
}

#[async_trait]
impl SecretStorage for CompositeSecretStorage {
    fn storage_type(&self) -> StorageBackend {
        StorageBackend::Memory // Composite doesn't have a specific type
    }

    async fn initialize(&self) -> SecretsResult<()> {
        self.global_storage.initialize().await?;
        self.world_storage.initialize().await?;
        self.user_storage.initialize().await?;
        Ok(())
    }

    async fn exists(&self, key: &str, context: &SecretContext) -> SecretsResult<bool> {
        self.storage_for(context).exists(key, context).await
    }

    async fn get(&self, key: &str, context: &SecretContext) -> SecretsResult<Option<serde_json::Value>> {
        self.storage_for(context).get(key, context).await
    }

    async fn set(
        &self,
        key: &str,
        value: serde_json::Value,
        context: &SecretContext,
        config: Option<SecretConfig>,
    ) -> SecretsResult<bool> {
        self.storage_for(context).set(key, value, context, config).await
    }

    async fn delete(&self, key: &str, context: &SecretContext) -> SecretsResult<bool> {
        self.storage_for(context).delete(key, context).await
    }

    async fn list(&self, context: &SecretContext) -> SecretsResult<SecretMetadata> {
        self.storage_for(context).list(context).await
    }

    async fn get_config(&self, key: &str, context: &SecretContext) -> SecretsResult<Option<SecretConfig>> {
        self.storage_for(context).get_config(key, context).await
    }

    async fn update_config(
        &self,
        key: &str,
        context: &SecretContext,
        config: SecretConfig,
    ) -> SecretsResult<bool> {
        self.storage_for(context).update_config(key, context, config).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_memory_storage_basic() {
        let storage = MemorySecretStorage::new();
        storage.initialize().await.unwrap();

        let context = SecretContext::global("agent-123");

        // Initially not exists
        assert!(!storage.exists("TEST_KEY", &context).await.unwrap());

        // Set a value
        let value = serde_json::json!("secret-value");
        storage.set("TEST_KEY", value.clone(), &context, None).await.unwrap();

        // Now exists
        assert!(storage.exists("TEST_KEY", &context).await.unwrap());

        // Get the value
        let retrieved = storage.get("TEST_KEY", &context).await.unwrap();
        assert_eq!(retrieved, Some(value));

        // Delete
        assert!(storage.delete("TEST_KEY", &context).await.unwrap());
        assert!(!storage.exists("TEST_KEY", &context).await.unwrap());
    }

    #[tokio::test]
    async fn test_memory_storage_list() {
        let storage = MemorySecretStorage::new();
        storage.initialize().await.unwrap();

        let context = SecretContext::global("agent-123");

        storage.set("KEY1", serde_json::json!("value1"), &context, None).await.unwrap();
        storage.set("KEY2", serde_json::json!("value2"), &context, None).await.unwrap();

        let metadata = storage.list(&context).await.unwrap();
        assert_eq!(metadata.count, 2);
        assert!(metadata.keys.contains(&"KEY1".to_string()));
        assert!(metadata.keys.contains(&"KEY2".to_string()));
    }

    #[tokio::test]
    async fn test_memory_storage_config() {
        let storage = MemorySecretStorage::new();
        storage.initialize().await.unwrap();

        let context = SecretContext::global("agent-123");

        let mut config = default_config();
        config.description = Some("Test secret".to_string());

        storage.set("TEST_KEY", serde_json::json!("value"), &context, Some(config.clone())).await.unwrap();

        let retrieved_config = storage.get_config("TEST_KEY", &context).await.unwrap();
        assert!(retrieved_config.is_some());
        assert_eq!(retrieved_config.unwrap().description, Some("Test secret".to_string()));
    }

    #[tokio::test]
    async fn test_composite_storage() {
        let storage = CompositeSecretStorage::with_memory_storage();
        storage.initialize().await.unwrap();

        let global_ctx = SecretContext::global("agent-123");
        let world_ctx = SecretContext::world("agent-123", "world-456");
        let user_ctx = SecretContext::user("agent-123", "user-789", None);

        // Set different values in different levels
        storage.set("KEY", serde_json::json!("global"), &global_ctx, None).await.unwrap();
        storage.set("KEY", serde_json::json!("world"), &world_ctx, None).await.unwrap();
        storage.set("KEY", serde_json::json!("user"), &user_ctx, None).await.unwrap();

        // Retrieve from each level
        assert_eq!(storage.get("KEY", &global_ctx).await.unwrap(), Some(serde_json::json!("global")));
        assert_eq!(storage.get("KEY", &world_ctx).await.unwrap(), Some(serde_json::json!("world")));
        assert_eq!(storage.get("KEY", &user_ctx).await.unwrap(), Some(serde_json::json!("user")));
    }

    #[tokio::test]
    async fn test_storage_key_generation() {
        let global_key = make_storage_key("TEST", &SecretContext::global("agent"));
        assert_eq!(global_key, "global:TEST");

        let world_key = make_storage_key("TEST", &SecretContext::world("agent", "world"));
        assert_eq!(world_key, "world:world:TEST");

        let user_key = make_storage_key("TEST", &SecretContext::user("agent", "user", None));
        assert_eq!(user_key, "user:user:TEST");
    }
}

File: plugins/plugin-secrets-manager/python/elizaos_plugin_secrets_manager/storage.py
"""
Storage implementations for secrets.

Provides in-memory storage and interfaces for elizaOS-native storage.
"""

from abc import ABC, abstractmethod
from typing import Optional, Dict
import time

from .types import (
    SecretConfig,
    SecretContext,
    SecretMetadata,
    SecretLevel,
    SecretType,
    SecretStatus,
    StorageBackend,
    StoredSecret,
)


class SecretStorageInterface(ABC):
    """Abstract interface for secret storage backends."""
    
    @property
    @abstractmethod
    def storage_type(self) -> StorageBackend:
        """Get the storage backend type."""
        pass
    
    @abstractmethod
    async def initialize(self) -> None:
        """Initialize the storage backend."""
        pass
    
    @abstractmethod
    async def exists(self, key: str, context: SecretContext) -> bool:
        """Check if a secret exists."""
        pass
    
    @abstractmethod
    async def get(self, key: str, context: SecretContext) -> Optional[str]:
        """Get a secret value."""
        pass
    
    @abstractmethod
    async def set(
        self,
        key: str,
        value: str,
        context: SecretContext,
        config: Optional[Dict] = None,
    ) -> bool:
        """Set a secret value."""
        pass
    
    @abstractmethod
    async def delete(self, key: str, context: SecretContext) -> bool:
        """Delete a secret."""
        pass
    
    @abstractmethod
    async def list(self, context: SecretContext) -> SecretMetadata:
        """List all secrets in a context."""
        pass
    
    @abstractmethod
    async def get_config(self, key: str, context: SecretContext) -> Optional[SecretConfig]:
        """Get secret configuration without value."""
        pass
    
    @abstractmethod
    async def update_config(
        self,
        key: str,
        context: SecretContext,
        config: Dict,
    ) -> bool:
        """Update secret configuration."""
        pass
    
    def _create_default_config(
        self,
        key: str,
        context: SecretContext,
        partial: Optional[Dict] = None,
    ) -> SecretConfig:
        """Create a default secret configuration."""
        partial = partial or {}
        
        return SecretConfig(
            type=SecretType(partial.get("type", "secret")),
            required=partial.get("required", False),
            description=partial.get("description", f"Secret: {key}"),
            can_generate=partial.get("canGenerate", False),
            validation_method=partial.get("validationMethod"),
            status=SecretStatus(partial.get("status", "valid")),
            last_error=partial.get("lastError"),
            attempts=partial.get("attempts", 0),
            created_at=partial.get("createdAt") or int(time.time() * 1000),
            validated_at=partial.get("validatedAt") or int(time.time() * 1000),
            plugin=partial.get("plugin", context.level.value),
            level=context.level,
            owner_id=context.user_id,
            world_id=context.world_id,
            encrypted=partial.get("encrypted", True),
            permissions=[],
            shared_with=[],
            expires_at=partial.get("expiresAt"),
        )


class MemorySecretStorage(SecretStorageInterface):
    """
    In-memory storage backend for secrets.
    
    Useful for testing and ephemeral environments.
    """
    
    def __init__(self):
        self._store: Dict[str, StoredSecret] = {}
    
    @property
    def storage_type(self) -> StorageBackend:
        return StorageBackend.MEMORY
    
    async def initialize(self) -> None:
        """Initialize - nothing to do for memory storage."""
        pass
    
    def _generate_storage_key(self, key: str, context: SecretContext) -> str:
        """Generate a unique storage key from the secret key and context."""
        if context.level == SecretLevel.GLOBAL:
            return f"global:{context.agent_id}:{key}"
        elif context.level == SecretLevel.WORLD:
            return f"world:{context.world_id}:{key}"
        elif context.level == SecretLevel.USER:
            return f"user:{context.user_id}:{key}"
        else:
            return f"unknown:{key}"
    
    def _get_context_prefix(self, context: SecretContext) -> str:
        """Get the storage key prefix for a context level."""
        if context.level == SecretLevel.GLOBAL:
            return f"global:{context.agent_id}:"
        elif context.level == SecretLevel.WORLD:
            return f"world:{context.world_id}:"
        elif context.level == SecretLevel.USER:
            return f"user:{context.user_id}:"
        else:
            return ""
    
    async def exists(self, key: str, context: SecretContext) -> bool:
        storage_key = self._generate_storage_key(key, context)
        return storage_key in self._store
    
    async def get(self, key: str, context: SecretContext) -> Optional[str]:
        storage_key = self._generate_storage_key(key, context)
        entry = self._store.get(storage_key)
        
        if not entry:
            return None
        
        # Check expiration
        if entry.config.expires_at and entry.config.expires_at < int(time.time() * 1000):
            del self._store[storage_key]
            return None
        
        # Return value (may be encrypted - caller handles decryption)
        if isinstance(entry.value, str):
            return entry.value
        
        # Return the encrypted object for caller to decrypt
        return entry.value
    
    async def set(
        self,
        key: str,
        value: str,
        context: SecretContext,
        config: Optional[Dict] = None,
    ) -> bool:
        storage_key = self._generate_storage_key(key, context)
        existing = self._store.get(storage_key)
        existing_config = existing.config if existing else None
        
        # Merge configs
        merged_config = {}
        if existing_config:
            merged_config = existing_config.to_dict()
        if config:
            merged_config.update(config)
        
        full_config = self._create_default_config(key, context, merged_config)
        
        self._store[storage_key] = StoredSecret(
            value=value,
            config=full_config,
        )
        
        return True
    
    async def delete(self, key: str, context: SecretContext) -> bool:
        storage_key = self._generate_storage_key(key, context)
        if storage_key in self._store:
            del self._store[storage_key]
            return True
        return False
    
    async def list(self, context: SecretContext) -> SecretMetadata:
        prefix = self._get_context_prefix(context)
        metadata: SecretMetadata = {}
        
        now = int(time.time() * 1000)
        keys_to_delete = []
        
        for storage_key, entry in self._store.items():
            if not storage_key.startswith(prefix):
                continue
            
            # Check expiration
            if entry.config.expires_at and entry.config.expires_at < now:
                keys_to_delete.append(storage_key)
                continue
            
            # Extract original key
            original_key = storage_key[len(prefix):]
            metadata[original_key] = entry.config
        
        # Clean up expired entries
        for key in keys_to_delete:
            del self._store[key]
        
        return metadata
    
    async def get_config(self, key: str, context: SecretContext) -> Optional[SecretConfig]:
        storage_key = self._generate_storage_key(key, context)
        entry = self._store.get(storage_key)
        
        if not entry:
            return None
        
        return entry.config
    
    async def update_config(
        self,
        key: str,
        context: SecretContext,
        config: Dict,
    ) -> bool:
        storage_key = self._generate_storage_key(key, context)
        entry = self._store.get(storage_key)
        
        if not entry:
            return False
        
        # Update config fields
        for field, value in config.items():
            if hasattr(entry.config, field):
                setattr(entry.config, field, value)
        
        return True
    
    def clear(self) -> None:
        """Clear all stored secrets (for testing)."""
        self._store.clear()
    
    def size(self) -> int:
        """Get the number of stored secrets."""
        return len(self._store)


class CompositeSecretStorage(SecretStorageInterface):
    """
    Composite storage that delegates to different backends based on context.
    """
    
    def __init__(
        self,
        global_storage: SecretStorageInterface,
        world_storage: SecretStorageInterface,
        user_storage: SecretStorageInterface,
    ):
        self._global_storage = global_storage
        self._world_storage = world_storage
        self._user_storage = user_storage
    
    @property
    def storage_type(self) -> StorageBackend:
        return StorageBackend.MEMORY
    
    async def initialize(self) -> None:
        await self._global_storage.initialize()
        await self._world_storage.initialize()
        await self._user_storage.initialize()
    
    def _get_storage_for_context(self, context: SecretContext) -> SecretStorageInterface:
        if context.level == SecretLevel.GLOBAL:
            return self._global_storage
        elif context.level == SecretLevel.WORLD:
            return self._world_storage
        elif context.level == SecretLevel.USER:
            return self._user_storage
        else:
            return self._global_storage
    
    async def exists(self, key: str, context: SecretContext) -> bool:
        storage = self._get_storage_for_context(context)
        return await storage.exists(key, context)
    
    async def get(self, key: str, context: SecretContext) -> Optional[str]:
        storage = self._get_storage_for_context(context)
        return await storage.get(key, context)
    
    async def set(
        self,
        key: str,
        value: str,
        context: SecretContext,
        config: Optional[Dict] = None,
    ) -> bool:
        storage = self._get_storage_for_context(context)
        return await storage.set(key, value, context, config)
    
    async def delete(self, key: str, context: SecretContext) -> bool:
        storage = self._get_storage_for_context(context)
        return await storage.delete(key, context)
    
    async def list(self, context: SecretContext) -> SecretMetadata:
        storage = self._get_storage_for_context(context)
        return await storage.list(context)
    
    async def get_config(self, key: str, context: SecretContext) -> Optional[SecretConfig]:
        storage = self._get_storage_for_context(context)
        return await storage.get_config(key, context)
    
    async def update_config(
        self,
        key: str,
        context: SecretContext,
        config: Dict,
    ) -> bool:
        storage = self._get_storage_for_context(context)
        return await storage.update_config(key, context, config)
